### DISCUSSION TITLE
Is it possible to define default prefixes with connection.prepareQuery(String)?

### DISCUSSION THREAD

**Discussion by ate47** [eclipse-rdf4j/rdf4j#3980] [Q&A] [answered]
Hello,

I was asking if it is possible to define default SPARQL prefixes without having to declare them in the query?

Like I want to be able to run the query

```sparql
SELECT * {ex:aaa ?p ?o}
```

without having to specify `PREFIX ex: <http://example.org/>`.

For now I was concatenating `PREFIX ex: <http://example.org/> ` to my query before running it, but  if someone try to define `ex:` it fails.

I've also tried `RepositoryConnection#setNamespace(String prefix, String name)`, but it doesn't seem to do anything.

EDIT:

To answer, I've added in the PR #3993 these methods to set custom prefixes

```java
// SPARQLParser
SPARQLParser(Set<Namespace> customDefaultPrefixes)

// SPARQLParserFactory
QueryParser getParser(Set<Namespace> customDefaultPrefixes);
```

Examples

```java
// create a set with the ex namespace 
Set<Namespace> customDefaultPrefixes = new HashSet<>();
customDefaultPrefixes.add(Values.namespace("ex", "http://example.org/"));

// create a parser with the default prefixes
SPARQLParser parser = new SPARQLParser(customDefaultPrefixes);

// we can use the default prefix without having to specify it
parser.parseQuery("SELECT ?s {?s ex:aaa ex:ooo}", null);

// we can redefine the prefix in the query if we want another one (this is not an error)
parser.parseQuery("PREFIX ex: <http://example2.org/> SELECT ?s {?s ex:aaa ex:ooo}", null);

// we cannot redefine a prefix in the same request (this is still an excepted error)
parser.parseQuery("PREFIX ex: <http://example.org/> PREFIX ex: <http://example2.org/> SELECT ?s {?s ex:aaa ex:ooo}", null);
```

**Comment by EmptyStar** [comment, marked as answer]
I came across this question a couple days ago in search of a solution to this exact situation, and from what I can tell, there is no API to apply prefixes to a SPARQL query or parser. It is possible for RDF parsers by calling `RDFParser.getParserConfig().set(BasicParserSettings.NAMESPACES,namespaces)`, but no analogue exists for SPARQL parsers that I could find.

I implemented it in my own project by lexing SPARQL queries for their prefixes and tacking on prefixes that don't exist which is the same approach you mentioned plus some logic to work around duplicate prefixes. It's not ideal, but it works. https://github.com/rdfpub/generator/blob/main/init/src/main/java/pub/rdf/builder/DatabaseBuilder.java#L163-L183

As far as `RepositoryConnection#setNamespace(String prefix, String name)` goes, I don't think those prefixes are used for much, if anything, other than making those prefixes available to be fetched by clients either via Java API or REST API. I think the RDF4J Workbench uses them to pre-populate SPARQL query text boxes, but I don't think they actually get applied to anything under the hood. But I'm not 100% on this and it's been a while since I've tinkered with the Workbench; feel free to correct me if I'm wrong.

It would be nice to have an actual API to apply prefixes to SPARQL queries.

**Reply by ate47** [reply]
I've made this pull request to add something similar to the API #3993, but if it doesn't pass, I will probably use your method, thanks.
